### PR TITLE
Add stop sequence to text generation pipeline

### PR DESCRIPTION
**Pull request by KMFODA**
# What does this PR do?

As per the conversation in https://github.com/huggingface/transformers/issues/17562, creating this draft PR to add a stop_sequence option to text generation pipelines.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@Narsil

Models:

All

Library:

- text generation: @patrickvonplaten
- pipelines: @LysandreJik

### CODE REVIEW

**Comment by KMFODA** [comment]
Hey @Narsil. I've managed to get this working for greedy decoding and multimodal sampling. For beam-search, what would be the best approach to deal with a stop_sequence? I've assumed that if a stop_sequence appears in any of the beams then we stop the generation process.

Should it instead be that we wait until each beam reaches the stop_sequence or any other stopping criteria before stopping the generation process?

File: src/transformers/pipelines/text2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/text2text_generation.py:84]: Why `3` ?

I think you're looking for `tokenizer.encode(stop_sequence, add_special_tokens=False)` which should raise a warning if its length is `>1` .

Not all tokenizer insert BOS and EOS, that's why we have to do it this way.

File: src/transformers/pipelines/text2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/text2text_generation.py:85]: Great message.

File: src/transformers/pipelines/text2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/text2text_generation.py:86]: `[0]` with the fix above.

File: src/transformers/pipelines/text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/text_generation.py:118]: I think implementing either in `generate` or here is enough.
We shouldn't try to implement it everywhere.

@gante @patrickvonplaten Are you ok if it's directly included in `generate` (otherwise we can keep it just in the pipeline).

File: src/transformers/pipelines/text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/text_generation.py:216]: Typo I think ?

File: src/transformers/generation_stopping_criteria.py
Comment (Narsil) [inline src/transformers/generation_stopping_criteria.py:124]: I think this line is wrong and the following one is correct, no ?

**Review by Narsil — COMMENTED** [review]
LGTM.

I pinged other maintainers to get an advice.

The main thing is that EOS is already handled without a stopping criteria so I don't know if we should add the new `StoppingCriteria`.

Also we should add some simple tests.

Ideally just set up a random model from hf-internal-testing, generate 5 tokens, look at the results and use token `3` as the new eos_token_id, decode it to get it as string, then reuse the generation with `generate(..., stop_sequence='xx')` and verify we stopped at token 3.

(We can leave the first steps with checks just so that the readers of the test can understand why we're supposed to stop at token `3`).

**Comment by Narsil** [comment]
> Should it instead be that we wait until each beam reaches the stop_sequence or any other stopping criteria before stopping the generation process?

@KMFODA I think `eos_token_id` is already handled for beam search, see my comment on the `StoppingCriteria`.

I will let others comment on the best way to do this in `.generate` but I think we don't need the criteria, just let `eos_token_id` regular logic apply (it's handled separately from `StoppingCriteria`).

**Comment by Narsil** [comment]
For the tests removing the breakpoint should help then for code quality.

```
pip install -e .[quality]
make fixup
```
Should do the trick.

**Comment by gante** [comment]
@Narsil @KMFODA I'm in favor of moving it to a `StoppingCriteria`, so that all conditions that can terminate generation fall under the same class. However, it should be noted that it is not a requirement to complete the issue, i.e. to add a stop sequence to the text generation pipeline :P 

It is already implemented on the multiple generation strategies (e.g. [here](https://github.com/huggingface/transformers/blob/main/src/transformers/generation_utils.py#L1744) for greedy search). Also, the existing implementation is different from the current PR -- the existing implementation only checks whether the `eos_token` is present in newly generated tokens. This is because models like `GPT-2` often set `pad_token_id` to `eos_token_id`, and we don't want the pad tokens to trigger this condition.

File: src/transformers/generation_stopping_criteria.py
Comment (KMFODA) [reply src/transformers/generation_stopping_criteria.py:124]: Hey @Narsil thanks for the helpful comments. And apologies for the few quality errors I was planning on addressing those after we decided on the strategy for eos_token_id. 

Here specifically I had 2 different returns because I was experimenting with different approaches to stop a beam search. One stops if any of the beams reach the eos_token_id the other waits for all of them to reach the eos_token_id. I'll remove both for now and address adding eos_token_id as a stopping criteria in a separate PR as per @gante's comments.

**Comment by KMFODA** [comment]
Thanks @Narsil @gante. Okay so for the sake of deploying iteratively I've removed the `eos_token_id` from the `StoppingCriteria` and will add it as a separate PR.

I've added a test for the `stop_sequence` being fed in at the pipeline level. When @Narsil's comment around wether the stop sequence should be handled in the `pipeline` or in the `generation_kwargs` is addressed I can alter this test accordingly.

File: src/transformers/generation_stopping_criteria.py
Comment (Narsil) [reply src/transformers/generation_stopping_criteria.py:124]: Oh I see. I think beam search should wait on all beams being "done". (And you keep the eos ones has long as their score allows.

I didn't touch enough the beam search code to be sure how to handle that.
Also I think you only need to check then final tokens to prevent looping over the whole `input_ids`.

`input_ids[:, -1] == eos_token_id` no ?

File: tests/generation/test_generation_utils.py
Comment (Narsil) [inline tests/generation/test_generation_utils.py:1985]:
```suggestion
    def test_stop_sequence_stopping_criteria(self):
        prompt = """Hello I believe in"""
        generator = pipeline("text-generation", model="hf-internal-testing/tiny-random-bart")
        output = generator(prompt)
        self.assertEqual(output, [{'generated_text': 'Hello I believe in in in number number number number number number number number number'}])
        
        output = generator(prompt, stop_sequence=" number")
        self.assertEqual(output, [{'generated_text': 'Hello I believe in in in number'}])
```

I think this formulation conveys the intent of the test a tiny bit better.
Also since you were only testing the last generated token, if we deactivated the whole option your test would still pass since the model just generates `number` all the time.
Wdyt?

**Review by Narsil — COMMENTED** [review]
This looks good to me. 

I think if the test needs to change if it's going to be in `generate` testing files. to only use `generate`.

We should implement `stop_sequence` only once (probably in `generate`) but we could have 2 tests if you want to test the full pipeline too. (Probably in `tests/pipelines/test_pipelines_text_generation.py` for instance.)

File: src/transformers/generation_stopping_criteria.py
Comment (gante) [reply src/transformers/generation_stopping_criteria.py:124]: It's a bit more complex than that -- we can only check new tokens, but different batch sequences may generate `eos_token_id` at different times.

The generate functions use an auxiliary variable to keep track of which members of the batch have already finished, see [here](https://github.com/huggingface/transformers/blob/main/src/transformers/generation_utils.py#L1744)

File: tests/generation/test_generation_utils.py
Comment (KMFODA) [reply tests/generation/test_generation_utils.py:1985]: this makes sense. I'll change it to that.

**Comment by KMFODA** [comment]
> We should implement `stop_sequence` only once (probably in `generate`) but we could have 2 tests if you want to test the full pipeline too. (Probably in `tests/pipelines/test_pipelines_text_generation.py` for instance.)

If we were to move `stop_sequence` to be in `generate` wouldn't we have to tokenise it first. In that case what's the reasoning behind feeding it as a `stop_sequence` instead of a `eos_token_id`?

**Comment by Narsil** [comment]
> If we were to move stop_sequence to be in generate wouldn't we have to tokenise it first. In that case what's the reasoning behind feeding it as a stop_sequence instead of a eos_token_id?

You're entirely right, oversight on my part. `eos_token_id` already does the job. So we just need to implement `stop_sequence` in the pipeline to tokenize the `stop_sequence` and produce the `eos_token_id` and just feed it to generate.
So no additional code in `generate` should be needed actually.

Sorry, failed to see that.

**Comment by KMFODA** [comment]
No problem I've just moved the stop_sequence back to the pipeline function and added the tests you requested in the `tests/pipelines/test_pipelines_text_generation.py` folder. This should make this PR ready for review now.

When I was playing with the stop_sequence though I found that sometime when I add a specific stop_sequence the output changes and avoids mentioning the word entirely. I don't have live examples now but I just wanted to check if this is normal behaviour? If not I can find examples on public models and share it in a different issue.

File: src/transformers/generation_utils.py
Comment (Narsil) [inline src/transformers/generation_utils.py:1066]: I think without an `s` is actually better, no ? There's only a single decay.

**Review by Narsil — APPROVED** [review]
LGTM  !

File: tests/pipelines/test_pipelines_text_generation.py
Comment (gante) [reply tests/pipelines/test_pipelines_text_generation.py:152]: ^ +1

**Review by gante — APPROVED** [review]
Other than the two comments I added and the failing tests, LGTM as well 👍

**Comment by patrickvonplaten** [comment]
@KMFODA I think your PR is almost ready to be merged! Would you like to try to fix the final problems and apply the review suggestions? :-)

**Comment by KMFODA** [comment]
Hey @patrickvonplaten. My apologies I was out sick over the past month. I worked on the suggestions now. Hopefully this should be good to merge now but if not let me know!

**Review by patrickvonplaten — APPROVED** [review]
Very nice! @gante I let you merge the PR :-)

**Comment by gante** [comment]
I'm happy with the PR, except for the `EndOfStringCriteria` class -- it is not being used, and it is not a good practice to add unused classes/functions. 

@KMFODA can you remove it for now, and perhaps reintroduce it in a follow-up PR (with use cases)? :)

**Comment by KMFODA** [comment]
Hi @gante yes of course. I had removed it locally but somehow the changes didn't push through with one of the commits. Forced changed it now. Hopefully that looks good now :).

**Pull request merged**
